# Tactical Asset Allocation - Group Assignment

# Importation of the different modules:

In [1]:
import yfinance as yf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Importation of Data from the S&P500

In [2]:
URL = 'https://en.wikipedia.org/wiki/List_of_S%26P_500_companies' # Wikipedia Webpage

L_tickers = pd.read_html(URL)[0]['Symbol'].tolist() #List Containing the Ticker names

# Define start and end dates
start_date = '2000-01-01'
end_date = '2022-12-31'

# Fetch adjusted close prices for the selected stocks
stock_data = yf.download(L_tickers, start=start_date, end=end_date)['Adj Close']


stock_data

[*********************100%%**********************]  503 of 503 completed


4 Failed downloads:
['BF.B']: Exception('%ticker%: No price data found, symbol may be delisted (1d 2000-01-01 -> 2022-12-31)')
['VLTO', 'KVUE']: Exception("%ticker%: Data doesn't exist for startDate = 946702800, endDate = 1672462800")
['BRK.B']: Exception('%ticker%: No timezone found, symbol may be delisted')


A    AAL        AAPL        ABBV       ABNB         ABT  \
Date                                                                           
2000-01-03   43.757751    NaN    0.847207         NaN        NaN    9.081109   
2000-01-04   40.415142    NaN    0.775779         NaN        NaN    8.821645   
2000-01-05   37.908188    NaN    0.787131         NaN        NaN    8.805433   
2000-01-06   36.464790    NaN    0.719014         NaN        NaN    9.113543   
2000-01-07   39.503521    NaN    0.753073         NaN        NaN    9.210837   
...                ...    ...         ...         ...        ...         ...   
2022-12-23  148.183105  12.71  131.127060  156.766937  85.250000  106.058319   
2022-12-27  148.500885  12.53  129.307251  156.661194  83.489998  106.440666   
2022-12-28  147.051102  12.32  125.339409  155.930710  82.489998  105.715187   
2022-12-29  150.030060  12.70  128.889587  156.247879  85.230003  108.146538   
2022-12-30  148.821777  12.72  129.207779  155.334778  85.500000  107.636726   

                 ACGL         ACN        ADBE         ADI  ...       WYNN  \
Date                                                       ...              
2000-01-03   1.277778         NaN   16.274670   28.572298  ...        NaN   
2000-01-04   1.270833         NaN   14.909399   27.126852  ...        NaN   
2000-01-05   1.388889         NaN   15.204174   27.522865  ...        NaN   
2000-01-06   1.375000         NaN   15.328288   26.770443  ...        NaN   
2000-01-07   1.451389         NaN   16.072981   27.522865  ...        NaN   
...               ...         ...         ...         ...  ...        ...   
2022-12-23  63.380001  261.928558  338.450012  160.778229  ...  80.091263   
2022-12-27  63.619999  261.160736  335.089996  159.168869  ...  83.673157   
2022-12-28  62.599998  259.004974  328.329987  157.284744  ...  79.466171   
2022-12-29  63.110001  264.182739  337.579987  160.915619  ...  80.627068   
2022-12-30  62.779999  262.666779  336.529999  160.964676  ...  81.827637   

                  XEL         XOM       XRAY         XYL         YUM  \
Date                                                                   
2000-01-03   7.112457   18.499041   6.695474         NaN    4.703721   
2000-01-04   7.276234   18.144711   6.695474         NaN    4.609174   
2000-01-05   7.556987   19.133877   6.811221         NaN    4.632810   
2000-01-06   7.486798   20.123060   6.820126         NaN    4.593416   
2000-01-07   7.486798   20.064001   6.802317         NaN    4.490990   
...               ...         ...        ...         ...         ...   
2022-12-23  68.689735  105.091942  31.343018  108.321205  126.519440   
2022-12-27  69.309525  106.552094  31.579344  109.298485  127.500984   
2022-12-28  68.810287  104.801842  30.506020  107.541344  126.921875   
2022-12-29  69.297813  105.594772  31.914906  110.206680  127.589325   
2022-12-30  68.361755  106.658463  31.479883  109.150406  125.714592   

                   ZBH        ZBRA       ZION         ZTS  
Date                                                       
2000-01-03         NaN   25.027779  36.146149         NaN  
2000-01-04         NaN   24.666668  34.395824         NaN  
2000-01-05         NaN   25.138889  34.355133         NaN  
2000-01-06         NaN   23.777779  34.843605         NaN  
2000-01-07         NaN   23.513889  34.925003         NaN  
...                ...         ...        ...         ...  
2022-12-23  125.965240  248.220001  46.140099  144.442429  
2022-12-27  126.551857  251.000000  46.511505  143.986572  
2022-12-28  125.269241  246.839996  45.682980  142.529861  
2022-12-29  127.098709  257.529999  46.740059  146.810791  
2022-12-30  126.770607  256.410004  46.816246  145.225266  

[5787 rows x 503 columns]

In [3]:
stocks_prices=stock_data
stocks_prices.dropna(axis=1, how='all', inplace=True) # Drop the columns that have only “NaN” entries.

In [4]:
# Drop also the company names that have more than 100 missing observations.
for stock in stocks_prices.columns:
    s=stocks_prices[stock].isna().sum() # Count the number of NaN
    if s>100:
        stocks_prices=stocks_prices.drop(columns=stock)
stocks_prices

A        AAPL         ABT       ACGL        ADBE  \
Date                                                                    
2000-01-03   43.757751    0.847207    9.081109   1.277778   16.274670   
2000-01-04   40.415142    0.775779    8.821645   1.270833   14.909399   
2000-01-05   37.908188    0.787131    8.805433   1.388889   15.204174   
2000-01-06   36.464790    0.719014    9.113543   1.375000   15.328288   
2000-01-07   39.503521    0.753073    9.210837   1.451389   16.072981   
...                ...         ...         ...        ...         ...   
2022-12-23  148.183105  131.127060  106.058319  63.380001  338.450012   
2022-12-27  148.500885  129.307251  106.440666  63.619999  335.089996   
2022-12-28  147.051102  125.339409  105.715187  62.599998  328.329987   
2022-12-29  150.030060  128.889587  108.146538  63.110001  337.579987   
2022-12-30  148.821777  129.207779  107.636726  62.779999  336.529999   

                   ADI        ADM         ADP        ADSK        AEE  ...  \
Date                                                                  ...   
2000-01-03   28.572298   6.473891   24.924681    8.052906  10.770641  ...   
2000-01-04   27.126852   6.406456   24.924681    7.660817  10.770641  ...   
2000-01-05   27.522865   6.305306   24.685011    7.178246  11.187304  ...   
2000-01-06   26.770443   6.339018   25.014557    6.740915  11.145635  ...   
2000-01-07   27.522865   6.440173   25.583738    7.540174  11.312300  ...   
...                ...        ...         ...         ...        ...  ...   
2022-12-23  160.778229  92.479225  235.585602  188.160004  86.777031  ...   
2022-12-27  159.168869  93.740044  235.673584  186.289993  87.503967  ...   
2022-12-28  157.284744  91.501862  232.564270  181.899994  86.631638  ...   
2022-12-29  160.915619  91.013168  235.272690  188.119995  87.242271  ...   
2022-12-30  160.964676  90.749275  233.551819  186.869995  86.185776  ...   

                   WMT        WRB         WST         WY        XEL  \
Date                                                                  
2000-01-03   43.555939   1.227682    5.431209  11.621564   7.112457   
2000-01-04   41.926144   1.189907    5.487316  11.184974   7.276234   
2000-01-05   41.070530   1.171020    5.464871  11.777491   7.556987   
2000-01-06   41.518703   1.239015    5.408766  12.328424   7.486798   
2000-01-07   44.656036   1.220127    5.464871  11.923014   7.486798   
...                ...        ...         ...        ...        ...   
2022-12-23  141.610062  71.400978  234.581253  29.921173  68.689735   
2022-12-27  141.649445  71.848160  236.177505  29.902166  69.309525   
2022-12-28  139.167313  70.769119  230.700333  29.104015  68.810287   
2022-12-29  140.014389  71.196838  238.422256  29.759640  69.297813   
2022-12-30  139.659790  70.545532  234.800720  29.455584  68.361755   

                   XOM       XRAY         YUM        ZBRA       ZION  
Date                                                                  
2000-01-03   18.499041   6.695474    4.703721   25.027779  36.146149  
2000-01-04   18.144711   6.695474    4.609174   24.666668  34.395824  
2000-01-05   19.133877   6.811221    4.632810   25.138889  34.355133  
2000-01-06   20.123060   6.820126    4.593416   23.777779  34.843605  
2000-01-07   20.064001   6.802317    4.490990   23.513889  34.925003  
...                ...        ...         ...         ...        ...  
2022-12-23  105.091942  31.343018  126.519440  248.220001  46.140099  
2022-12-27  106.552094  31.579344  127.500984  251.000000  46.511505  
2022-12-28  104.801842  30.506020  126.921875  246.839996  45.682980  
2022-12-29  105.594772  31.914906  127.589325  257.529999  46.740059  
2022-12-30  106.658463  31.479883  125.714592  256.410004  46.816246  

[5787 rows x 363 columns]

# Computation of returns

In [5]:
return_df=stocks_prices.pct_change()
return_df

A      AAPL       ABT      ACGL      ADBE       ADI  \
Date                                                                     
2000-01-03       NaN       NaN       NaN       NaN       NaN       NaN   
2000-01-04 -0.076389 -0.084310 -0.028572 -0.005435 -0.083889 -0.050589   
2000-01-05 -0.062030  0.014633 -0.001838  0.092897  0.019771  0.014599   
2000-01-06 -0.038076 -0.086538  0.034991 -0.010000  0.008163 -0.027338   
2000-01-07  0.083333  0.047369  0.010676  0.055556  0.048583  0.028106   
...              ...       ...       ...       ...       ...       ...   
2022-12-23  0.001476 -0.002798  0.001389  0.008433  0.005735  0.000550   
2022-12-27  0.002145 -0.013878  0.003605  0.003787 -0.009928 -0.010010   
2022-12-28 -0.009763 -0.030685 -0.006816 -0.016033 -0.020174 -0.011837   
2022-12-29  0.020258  0.028325  0.022999  0.008147  0.028173  0.023085   
2022-12-30 -0.008054  0.002469 -0.004714 -0.005229 -0.003110  0.000305   

                 ADM       ADP      ADSK       AEE  ...       WMT       WRB  \
Date                                                ...                       
2000-01-03       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2000-01-04 -0.010416  0.000000 -0.048689  0.000000  ... -0.037418 -0.030770   
2000-01-05 -0.015789 -0.009616 -0.062992  0.038685  ... -0.020408 -0.015873   
2000-01-06  0.005347  0.013350 -0.060924 -0.003725  ...  0.010912  0.058065   
2000-01-07  0.015957  0.022754  0.118568  0.014953  ...  0.075564 -0.015244   
...              ...       ...       ...       ...  ...       ...       ...   
2022-12-23  0.012087  0.007443  0.000319  0.013241  ...  0.002021  0.006440   
2022-12-27  0.013634  0.000373 -0.009938  0.008377  ...  0.000278  0.006263   
2022-12-28 -0.023876 -0.013193 -0.023565 -0.009969  ... -0.017523 -0.015018   
2022-12-29 -0.005341  0.011646  0.034195  0.007049  ...  0.006087  0.006044   
2022-12-30 -0.002900 -0.007314 -0.006645 -0.012110  ... -0.002533 -0.009148   

                 WST        WY       XEL       XOM      XRAY       YUM  \
Date                                                                     
2000-01-03       NaN       NaN       NaN       NaN       NaN       NaN   
2000-01-04  0.010331 -0.037567  0.023027 -0.019154  0.000000 -0.020101   
2000-01-05 -0.004090  0.052974  0.038585  0.054515  0.017287  0.005128   
2000-01-06 -0.010267  0.046779 -0.009288  0.051698  0.001307 -0.008503   
2000-01-07  0.010373 -0.032884  0.000000 -0.002935 -0.002611 -0.022298   
...              ...       ...       ...       ...       ...       ...   
2022-12-23 -0.000553  0.014171  0.012852  0.026445  0.011118  0.000621   
2022-12-27  0.006805 -0.000635  0.009023  0.013894  0.007540  0.007758   
2022-12-28 -0.023191 -0.026692 -0.007203 -0.016426 -0.033988 -0.004542   
2022-12-29  0.033472  0.022527  0.007085  0.007566  0.046184  0.005259   
2022-12-30 -0.015190 -0.010217 -0.013508  0.010073 -0.013631 -0.014693   

                ZBRA      ZION  
Date                            
2000-01-03       NaN       NaN  
2000-01-04 -0.014428 -0.048424  
2000-01-05  0.019144 -0.001183  
2000-01-06 -0.054144  0.014218  
2000-01-07 -0.011098  0.002336  
...              ...       ...  
2022-12-23  0.002869  0.003521  
2022-12-27  0.011200  0.008050  
2022-12-28 -0.016574 -0.017813  
2022-12-29  0.043307  0.023139  
2022-12-30 -0.004349  0.001630  

[5787 rows x 363 columns]

In [6]:
stocks_prices.index

Index([2000-01-03 00:00:00, 2000-01-04 00:00:00, 2000-01-05 00:00:00,
       2000-01-06 00:00:00, 2000-01-07 00:00:00, 2000-01-10 00:00:00,
       2000-01-11 00:00:00, 2000-01-12 00:00:00, 2000-01-13 00:00:00,
       2000-01-14 00:00:00,
       ...
       2022-12-16 00:00:00, 2022-12-19 00:00:00, 2022-12-20 00:00:00,
       2022-12-21 00:00:00, 2022-12-22 00:00:00, 2022-12-23 00:00:00,
       2022-12-27 00:00:00, 2022-12-28 00:00:00, 2022-12-29 00:00:00,
       2022-12-30 00:00:00],
      dtype='object', name='Date', length=5787)

In [7]:
#We create the monthly returns

stocks_prices.index = pd.to_datetime(stocks_prices.index)
month_stock_prices = stocks_prices.resample("1M").last()
month_stock_prices.index = month_stock_prices.index.date

month_return=month_stock_prices.pct_change()
month_return

A      AAPL       ABT      ACGL      ADBE       ADI  \
2000-01-31       NaN       NaN       NaN       NaN       NaN       NaN   
2000-02-29  0.569405  0.104820  0.013436  0.042735  0.852441  0.681818   
2000-03-31  0.001203  0.184842  0.066288  0.073771  0.091566  0.023847   
2000-04-30 -0.147836 -0.086517  0.097843 -0.060115  0.086468 -0.045807   
2000-05-31 -0.169252 -0.322922  0.058536 -0.025380 -0.069251  0.002441   
...              ...       ...       ...       ...       ...       ...   
2022-08-31 -0.043624 -0.031208 -0.056872  0.029730 -0.089437 -0.114566   
2022-09-30 -0.052242 -0.120977 -0.057379 -0.003937 -0.263068 -0.080446   
2022-10-31  0.140185  0.109551  0.027355  0.262626  0.157340  0.023540   
2022-11-30  0.120202 -0.033028  0.087326  0.041913  0.082983  0.205371   
2022-12-31 -0.032952 -0.122273  0.020543  0.047905 -0.024353 -0.041590   

                 ADM       ADP      ADSK       AEE  ...       WMT       WRB  \
2000-01-31       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2000-02-29 -0.140034 -0.081686  0.462168 -0.078695  ... -0.109589 -0.075540   
2000-03-31  0.024845  0.109850  0.018182  0.055017  ...  0.160432  0.447560   
2000-04-30 -0.036364  0.115285 -0.155447  0.185859  ... -0.019911 -0.084011   
2000-05-31  0.207110  0.022067 -0.030945  0.000000  ...  0.040632  0.044379   
...              ...       ...       ...       ...  ...       ...       ...   
2022-08-31  0.066813  0.013645 -0.067400 -0.005369  ...  0.008158  0.036303   
2022-09-30 -0.084651 -0.070548 -0.074056 -0.124761  ... -0.021501 -0.001846   
2022-10-31  0.205469  0.068571  0.147216  0.012042  ...  0.097379  0.151750   
2022-11-30  0.009645  0.092842 -0.057630  0.095682  ...  0.070891  0.025545   
2022-12-31 -0.047692 -0.091325 -0.074672  0.002299  ... -0.066235 -0.047314   

                 WST        WY       XEL       XOM      XRAY       YUM  \
2000-01-31       NaN       NaN       NaN       NaN       NaN       NaN   
2000-02-29 -0.048485 -0.105665 -0.087662 -0.086158  0.035353 -0.069869   
2000-03-31 -0.140127  0.110841  0.131673  0.035685  0.109781  0.166667   
2000-04-30 -0.027590 -0.063597  0.116965 -0.004006  0.024229  0.090543   
2000-05-31 -0.038363 -0.063303  0.014326  0.078226  0.055915 -0.134686   
...              ...       ...       ...       ...       ...       ...   
2022-08-31 -0.136424 -0.059472  0.014621 -0.004243 -0.093750 -0.087764   
2022-09-30 -0.170582 -0.159506 -0.132408 -0.086620 -0.131233 -0.044049   
2022-10-31 -0.064938  0.082983  0.017344  0.269156  0.087125  0.111999   
2022-11-30  0.020676  0.057549  0.078482  0.012872 -0.018170  0.093020   
2022-12-31  0.002940 -0.047033  0.005288 -0.009341  0.056477 -0.004508   

                ZBRA      ZION  
2000-01-31       NaN       NaN  
2000-02-29  0.124604 -0.102537  
2000-03-31 -0.248826 -0.215548  
2000-04-30  0.140000 -0.003003  
2000-05-31 -0.157895  0.131084  
...              ...       ...  
2022-08-31 -0.156700  0.015860  
2022-09-30 -0.131382 -0.075777  
2022-10-31  0.080951  0.021235  
2022-11-30 -0.045689  0.005822  
2022-12-31 -0.051317 -0.051332  

[276 rows x 363 columns]

# Calculation of Volatility

In [8]:
# Here we add a columns Year in our dataframe
L_year =[]
L_month=[]
for date in return_df.index:
    L_year.append(date.year)
    L_month.append(date.month)
return_df['Year']=L_year
return_df['Month']=L_month
return_df['Month-Year']=pd.to_datetime(return_df[['Month','Year']].assign(DAY=1))
return_df

/var/folders/gw/h3bw6tzs4nb5dx1ml0qj1mzr0000gn/T/ipykernel_965/2687220330.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  return_df['Year']=L_year
/var/folders/gw/h3bw6tzs4nb5dx1ml0qj1mzr0000gn/T/ipykernel_965/2687220330.py:8: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  return_df['Month']=L_month
/var/folders/gw/h3bw6tzs4nb5dx1ml0qj1mzr0000gn/T/ipykernel_965/2687220330.py:9: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Con

A      AAPL       ABT      ACGL      ADBE       ADI  \
Date                                                                     
2000-01-03       NaN       NaN       NaN       NaN       NaN       NaN   
2000-01-04 -0.076389 -0.084310 -0.028572 -0.005435 -0.083889 -0.050589   
2000-01-05 -0.062030  0.014633 -0.001838  0.092897  0.019771  0.014599   
2000-01-06 -0.038076 -0.086538  0.034991 -0.010000  0.008163 -0.027338   
2000-01-07  0.083333  0.047369  0.010676  0.055556  0.048583  0.028106   
...              ...       ...       ...       ...       ...       ...   
2022-12-23  0.001476 -0.002798  0.001389  0.008433  0.005735  0.000550   
2022-12-27  0.002145 -0.013878  0.003605  0.003787 -0.009928 -0.010010   
2022-12-28 -0.009763 -0.030685 -0.006816 -0.016033 -0.020174 -0.011837   
2022-12-29  0.020258  0.028325  0.022999  0.008147  0.028173  0.023085   
2022-12-30 -0.008054  0.002469 -0.004714 -0.005229 -0.003110  0.000305   

                 ADM       ADP      ADSK       AEE  ...        WY       XEL  \
Date                                                ...                       
2000-01-03       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2000-01-04 -0.010416  0.000000 -0.048689  0.000000  ... -0.037567  0.023027   
2000-01-05 -0.015789 -0.009616 -0.062992  0.038685  ...  0.052974  0.038585   
2000-01-06  0.005347  0.013350 -0.060924 -0.003725  ...  0.046779 -0.009288   
2000-01-07  0.015957  0.022754  0.118568  0.014953  ... -0.032884  0.000000   
...              ...       ...       ...       ...  ...       ...       ...   
2022-12-23  0.012087  0.007443  0.000319  0.013241  ...  0.014171  0.012852   
2022-12-27  0.013634  0.000373 -0.009938  0.008377  ... -0.000635  0.009023   
2022-12-28 -0.023876 -0.013193 -0.023565 -0.009969  ... -0.026692 -0.007203   
2022-12-29 -0.005341  0.011646  0.034195  0.007049  ...  0.022527  0.007085   
2022-12-30 -0.002900 -0.007314 -0.006645 -0.012110  ... -0.010217 -0.013508   

                 XOM      XRAY       YUM      ZBRA      ZION  Year  Month  \
Date                                                                        
2000-01-03       NaN       NaN       NaN       NaN       NaN  2000      1   
2000-01-04 -0.019154  0.000000 -0.020101 -0.014428 -0.048424  2000      1   
2000-01-05  0.054515  0.017287  0.005128  0.019144 -0.001183  2000      1   
2000-01-06  0.051698  0.001307 -0.008503 -0.054144  0.014218  2000      1   
2000-01-07 -0.002935 -0.002611 -0.022298 -0.011098  0.002336  2000      1   
...              ...       ...       ...       ...       ...   ...    ...   
2022-12-23  0.026445  0.011118  0.000621  0.002869  0.003521  2022     12   
2022-12-27  0.013894  0.007540  0.007758  0.011200  0.008050  2022     12   
2022-12-28 -0.016426 -0.033988 -0.004542 -0.016574 -0.017813  2022     12   
2022-12-29  0.007566  0.046184  0.005259  0.043307  0.023139  2022     12   
2022-12-30  0.010073 -0.013631 -0.014693 -0.004349  0.001630  2022     12   

            Month-Year  
Date                    
2000-01-03  2000-01-01  
2000-01-04  2000-01-01  
2000-01-05  2000-01-01  
2000-01-06  2000-01-01  
2000-01-07  2000-01-01  
...                ...  
2022-12-23  2022-12-01  
2022-12-27  2022-12-01  
2022-12-28  2022-12-01  
2022-12-29  2022-12-01  
2022-12-30  2022-12-01  

[5787 rows x 366 columns]

In [51]:
L_date_unique=list(np.unique(return_df['Month-Year']))

df_monthly_volatility=pd.DataFrame(index=L_date_unique, columns=return_df.columns[:363])

for stock in return_df.columns[:363]:
    L_mean=[]
    L_std=[]
    L_sr=[]
    for date in L_date_unique:
        data= return_df[return_df['Month-Year']== date][stock] 
        std=data.std()*(21**0.5) # Volatility
        L_std.append(std)

    df_monthly_volatility[stock]=L_std

df_monthly_volatility # This dataframe will be composed of all the volatility per year and per stock

A      AAPL       ABT      ACGL      ADBE       ADI  \
2000-01-01  0.176785  0.253842  0.109386  0.147413  0.194556  0.196413   
2000-02-01  0.274698  0.154532  0.122485  0.139534  0.231197  0.216932   
2000-03-01  0.518904  0.240758  0.128821  0.051605  0.324719  0.332484   
2000-04-01  0.343793  0.229795  0.154043  0.109676  0.354779  0.295870   
2000-05-01  0.390838  0.163517  0.102473  0.089136  0.276043  0.248615   
...              ...       ...       ...       ...       ...       ...   
2022-08-01  0.106404  0.074393  0.055185  0.051904  0.094286  0.097450   
2022-09-01  0.078711  0.105795  0.066533  0.093691  0.189855  0.076524   
2022-10-01  0.102174  0.122291  0.111079  0.147546  0.091643  0.116062   
2022-11-01  0.129005  0.137399  0.071135  0.067940  0.163228  0.138588   
2022-12-01  0.065776  0.083289  0.062515  0.057503  0.086621  0.070758   

                 ADM       ADP      ADSK       AEE  ...       WMT       WRB  \
2000-01-01  0.105047  0.154760  0.216084  0.076847  ...  0.144902  0.153391   
2000-02-01  0.086177  0.159758  0.262177  0.076669  ...  0.211071  0.094810   
2000-03-01  0.160166  0.155587  0.160319  0.089331  ...  0.165902  0.190181   
2000-04-01  0.122340  0.135374  0.212502  0.087713  ...  0.175866  0.080388   
2000-05-01  0.114131  0.103433  0.209196  0.068547  ...  0.158984  0.086132   
...              ...       ...       ...       ...  ...       ...       ...   
2022-08-01  0.066687  0.052476  0.092819  0.041673  ...  0.080232  0.063365   
2022-09-01  0.096795  0.064667  0.090494  0.082771  ...  0.059479  0.072763   
2022-10-01  0.096454  0.083839  0.116205  0.097836  ...  0.061456  0.095622   
2022-11-01  0.057970  0.087573  0.204579  0.073732  ...  0.082396  0.076331   
2022-12-01  0.081462  0.075988  0.089741  0.051429  ...  0.046514  0.051248   

                 WST        WY       XEL       XOM      XRAY       YUM  \
2000-01-01  0.072734  0.142800  0.070274  0.119431  0.083256  0.099345   
2000-02-01  0.163154  0.136455  0.066558  0.106415  0.069033  0.168969   
2000-03-01  0.103527  0.172556  0.133602  0.139844  0.106478  0.236783   
2000-04-01  0.087318  0.126413  0.098542  0.103637  0.129599  0.149374   
2000-05-01  0.072387  0.093501  0.090179  0.063371  0.125678  0.133936   
...              ...       ...       ...       ...       ...       ...   
2022-08-01  0.068275  0.073580  0.040352  0.098438  0.076645  0.055525   
2022-09-01  0.111063  0.110495  0.087196  0.096117  0.101042  0.065722   
2022-10-01  0.193818  0.082487  0.103775  0.096752  0.090773  0.090883   
2022-11-01  0.174891  0.127922  0.073665  0.081782  0.163370  0.043972   
2022-12-01  0.107085  0.075031  0.048090  0.070925  0.102543  0.041941   

                ZBRA      ZION  
2000-01-01  0.163824  0.181820  
2000-02-01  0.231040  0.146304  
2000-03-01  0.267266  0.319104  
2000-04-01  0.271308  0.179879  
2000-05-01  0.224340  0.110039  
...              ...       ...  
2022-08-01  0.143067  0.066841  
2022-09-01  0.111373  0.082610  
2022-10-01  0.134402  0.135277  
2022-11-01  0.228614  0.128691  
2022-12-01  0.101201  0.099411  

[276 rows x 363 columns]

### Train / Test / Validation:

Next, we can move on to splitting the dataset into training, validation, and test sets.
The dataset contains 276 data points, which is approximately 23 years of trading data.

Given this:

- We'll use roughly 22 years (264 months) for training.
- The remaining data (~12 months) will be used for validation.

In [42]:
month_stock_prices.columns

Index(['A', 'AAPL', 'ABT', 'ACGL', 'ADBE', 'ADI', 'ADM', 'ADP', 'ADSK', 'AEE',
       ...
       'WMT', 'WRB', 'WST', 'WY', 'XEL', 'XOM', 'XRAY', 'YUM', 'ZBRA', 'ZION'],
      dtype='object', length=363)

In [57]:
for stock in df_monthly_volatility.columns:
    
    s=df_monthly_volatility[stock].isna().sum()
    if s!=0:
        df_monthly_volatility=df_monthly_volatility.drop(columns=stock)

In [28]:
from sklearn.model_selection import train_test_split
X = df_monthly_volatility  # Features (stock prices of each day)
y = df_monthly_volatility.shift(-1)  # Targets (next day prices)
X_train, X_test, y_train, y_test = train_test_split(X[:-1], y[:-1], test_size = 12/276, random_state = 42)

In [29]:
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(263, 359)
(263, 359)
(12, 359)
(12, 359)


In [31]:
from sklearn.metrics import accuracy_score, confusion_matrix, ConfusionMatrixDisplay, classification_report
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error

# Random Forest

- First test : 10 trees

In [32]:
models = {}
for stock in X.columns:
    
    rf_df = RandomForestRegressor(n_estimators=10, random_state=42)
    rf_df.fit(X_train,y_train[stock])
    pred = rf_df.predict(X_test)
    models[stock] = rf_df
    
rf_pred = {}
for stock, m in models.items():
    rf_pred[stock] = m.predict(X_test)

rf_pred_df = pd.DataFrame(rf_pred)



In [33]:
L_10=[]
for stock in X.columns: # Evaluate baseline model performance
    model = models[stock]
    predictions = model.predict(X_test)
    mse = mean_squared_error(y_test[stock], predictions)
    L_10.append(np.sqrt(mse))

- Second test : 20 trees

In [35]:
models = {}
for stock in X.columns: 
    rf_df = RandomForestRegressor(n_estimators=20, random_state=42)
    rf_df.fit(X_train,y_train[stock])
    pred = rf_df.predict(X_test)
    
    models[stock] = rf_df
rf_pred = {}
for stock, m in models.items():
    rf_pred[stock] = m.predict(X_test)

rf_pred_df = pd.DataFrame(rf_pred)

In [36]:
L_20=[]
for stock in X.columns: # Evaluate baseline model performance
    model = models[stock]
    predictions = model.predict(X_test)
    mse = mean_squared_error(y_test[stock], predictions)
    L_20.append(np.sqrt(mse))

- Third test : 50 trees

In [37]:
models = {}
for stock in X.columns:
    rf_df = RandomForestRegressor(n_estimators=50, random_state=42)
    rf_df.fit(X_train,y_train[stock])
    pred = rf_df.predict(X_test)
    models[stock] = rf_df
rf_pred = {}
for stock, m in models.items():
    rf_pred[stock] = m.predict(X_test)

rf_pred_df = pd.DataFrame(rf_pred)

In [38]:
L_50=[]
for stock in X.columns: # Evaluate baseline model performance
    model = models[stock]
    predictions = model.predict(X_test)
    mse = mean_squared_error(y_test[stock], predictions)
    L_50.append(np.sqrt(mse))

In [ ]:
for stock in rf_pred_df.columns:
    plt.figure(figsize = (20,7))
    plt.plot(rf_pred_df.index,X_test[stock], label = "Real Value")
    plt.plot(rf_pred_df.index,rf_pred_df[stock], label = "Random Forest prediction 50 Trees")
    plt.title("Comparison of real data estimator " + stock)
    plt.legend(loc='upper left')
    plt.show()

In [39]:
list_stock=list(X.columns)
df_comparaison=pd.DataFrame()
df_comparaison['Stock']= list_stock
df_comparaison['RMSE 10 Trees']=L_10
df_comparaison['RMSE 20 Trees']=L_20
df_comparaison['RMSE 50 Trees']=L_50
df_comparaison=df_comparaison.set_index('Stock')
df_comparaison

RMSE 10 Trees  RMSE 20 Trees  RMSE 50 Trees
Stock                                             
A           0.030919       0.035368       0.025833
AAPL        0.045060       0.037450       0.035191
ABT         0.026253       0.023630       0.025255
ACGL        0.061159       0.042336       0.031724
ADBE        0.064233       0.048935       0.046865
...              ...            ...            ...
XOM         0.026816       0.033930       0.026023
XRAY        0.026879       0.027355       0.027185
YUM         0.045270       0.039643       0.033890
ZBRA        0.069725       0.066581       0.063708
ZION        0.093185       0.081023       0.078232

[359 rows x 3 columns]

The best model is the one with 50 trees, we will continue with this one

# Predictions:

In [202]:
def prediction_function(df,n_period):
    prediction=[]
    dataset=df_monthly_volatility.copy()

    for i in range(n_period):
        prediction=[]
        for stock, m in models.items():
            prediction.append(m.predict(dataset.tail(1)))

        prediction=pd.DataFrame(prediction,index=df_monthly_volatility.columns)
        dataset=pd.concat([dataset,prediction.transpose()])
    return dataset


In [204]:
prediction_function(df_monthly_volatility,12)

A      AAPL       ABT      ACGL      ADBE  \
2000-01-01 00:00:00  0.176785  0.253842  0.109386  0.147413  0.194556   
2000-02-01 00:00:00  0.274698  0.154532  0.122485  0.139534  0.231197   
2000-03-01 00:00:00  0.518904  0.240758  0.128821  0.051605  0.324719   
2000-04-01 00:00:00  0.343793  0.229795  0.154043  0.109676  0.354779   
2000-05-01 00:00:00  0.390838  0.163517  0.102473  0.089136  0.276043   
...                       ...       ...       ...       ...       ...   
0                    0.054820  0.061510  0.043519  0.042238  0.067599   
0                    0.056112  0.061052  0.043428  0.041109  0.066562   
0                    0.055028  0.062190  0.040574  0.040321  0.065793   
0                    0.054986  0.062694  0.040336  0.038925  0.063857   
0                    0.055510  0.059162  0.041314  0.039914  0.064098   

                          ADI       ADM       ADP      ADSK       AEE  ...  \
2000-01-01 00:00:00  0.196413  0.105047  0.154760  0.216084  0.076847  ...   
2000-02-01 00:00:00  0.216932  0.086177  0.159758  0.262177  0.076669  ...   
2000-03-01 00:00:00  0.332484  0.160166  0.155587  0.160319  0.089331  ...   
2000-04-01 00:00:00  0.295870  0.122340  0.135374  0.212502  0.087713  ...   
2000-05-01 00:00:00  0.248615  0.114131  0.103433  0.209196  0.068547  ...   
...                       ...       ...       ...       ...       ...  ...   
0                    0.062748  0.056668  0.039267  0.069038  0.042957  ...   
0                    0.060604  0.057454  0.040196  0.071024  0.042931  ...   
0                    0.060371  0.058254  0.038966  0.072202  0.043033  ...   
0                    0.060957  0.056951  0.038810  0.073083  0.043602  ...   
0                    0.059286  0.056100  0.038503  0.073119  0.043047  ...   

                          WMT       WRB       WST        WY       XEL  \
2000-01-01 00:00:00  0.144902  0.153391  0.072734  0.142800  0.070274   
2000-02-01 00:00:00  0.211071  0.094810  0.163154  0.136455  0.066558   
2000-03-01 00:00:00  0.165902  0.190181  0.103527  0.172556  0.133602   
2000-04-01 00:00:00  0.175866  0.080388  0.087318  0.126413  0.098542   
2000-05-01 00:00:00  0.158984  0.086132  0.072387  0.093501  0.090179   
...                       ...       ...       ...       ...       ...   
0                    0.037513  0.045392  0.054843  0.061000  0.042467   
0                    0.037795  0.044961  0.054844  0.062263  0.041277   
0                    0.037464  0.043543  0.055956  0.060991  0.041858   
0                    0.037528  0.043066  0.054155  0.061220  0.040742   
0                    0.036862  0.044884  0.054940  0.061993  0.040945   

                          XOM      XRAY       YUM      ZBRA      ZION  
2000-01-01 00:00:00  0.119431  0.083256  0.099345  0.163824  0.181820  
2000-02-01 00:00:00  0.106415  0.069033  0.168969  0.231040  0.146304  
2000-03-01 00:00:00  0.139844  0.106478  0.236783  0.267266  0.319104  
2000-04-01 00:00:00  0.103637  0.129599  0.149374  0.271308  0.179879  
2000-05-01 00:00:00  0.063371  0.125678  0.133936  0.224340  0.110039  
...                       ...       ...       ...       ...       ...  
0                    0.046717  0.054327  0.051108  0.059829  0.075193  
0                    0.044406  0.054381  0.051670  0.061689  0.075945  
0                    0.044524  0.053193  0.053022  0.062717  0.074229  
0                    0.043490  0.053319  0.052885  0.059964  0.076132  
0                    0.043788  0.054001  0.052458  0.059539  0.071644  

[288 rows x 359 columns]

In [198]:
def portfolio_construction(prediction): #argument= dataset.tail(periodicité)
    portfolio_composition_weights = []
    portfolio_composition_names=[]
    ranking=prediction.rank(axis=1)
    stock_lowest_vol={}
    for stock in prediction.columns:
        if ranking[stock].iloc[0]<21.0:
            portfolio_composition_names.append(stock)
    s=0      
    for stock in portfolio_composition_names:
        s+=1/prediction[stock].iloc[0]
    for stock in portfolio_composition_names:  
        w=(1/prediction[stock].iloc[0])/s
        portfolio_composition_weights.append(w)

    df_portfolio=pd.DataFrame()
    df_portfolio['Composition']=portfolio_composition_names
    df_portfolio['Weight']=portfolio_composition_weights
    return df_portfolio
        
    

In [199]:
portfolio_construction(dataset.tail(1))


Composition    Weight
0          ADP  0.048060
1          CLX  0.051633
2          DTE  0.047653
3          DUK  0.055232
4           ED  0.049654
5          GIS  0.049508
6          JNJ  0.055041
7            K  0.051070
8          KMB  0.047979
9           KO  0.051948
10         MCD  0.047627
11         MKC  0.048526
12         PEP  0.055053
13          PG  0.048419
14          SO  0.049323
15         SRE  0.047599
16           T  0.049422
17         WEC  0.047294
18          WM  0.048761
19         WMT  0.050199

# Calculation of Metrics

In [139]:
def mean_squared_error(y_true, y_pred):

    if len(y_true) != len(y_pred):
        raise ValueError("Input arrays must have the same length")

    # Calculate the squared differences between true and predicted values
    squared_errors = [(true - pred) ** 2 for true, pred in zip(y_true, y_pred)]

    # Calculate the mean squared error
    mse = sum(squared_errors) / len(y_true)

    return mse

In [149]:
def sharpe_ratio(returns, risk_free_rate):

    mean_return = np.mean(returns)
    std_return = np.std(returns)

    # Calculate the excess return
    excess_return = mean_return - risk_free_rate

    # Calculate the Sharpe Ratio
    sharpe_ratio = excess_return / std_return

    return sharpe_ratio

In [150]:
def information_ratio(portfolio_returns, benchmark_returns):

    # Calculate the active return
    active_return = np.array(portfolio_returns) - np.array(benchmark_returns)

    # Calculate the tracking error (standard deviation of active return)
    tracking_error = np.std(active_return)

    # Calculate the Information Ratio
    information_ratio = np.mean(active_return) / tracking_error

    return information_ratio

In [151]:
def calculate_skewness(data):
 
    n = len(data)
    mean_data = np.mean(data)
    std_data = np.std(data)

    skewness = (n / ((n - 1) * (n - 2))) * np.sum(((data - mean_data) / std_data) ** 3)

    return skewness

In [152]:
def calculate_maximum_drawdown(prices):

    peak_value = np.maximum.accumulate(prices)
    trough_value = np.minimum.accumulate(prices)
    
    drawdowns = (peak_value - trough_value) / peak_value
    max_drawdown = np.max(drawdowns) * 100  # Convert to percentage

    return max_drawdown

In [153]:
def calculate_hit_ratio(profitable_trades, total_trades):

    if total_trades == 0:
        return 0.0

    hit_ratio = (profitable_trades / total_trades) * 100
    return hit_ratio

In [154]:
def calculate_turnover(buy_value, sell_value, average_portfolio_value):

    total_value_traded = buy_value + sell_value
    turnover_ratio = (total_value_traded / average_portfolio_value) * 100
    return turnover_ratio